src/App.tsx

```diff
@@ -15,48 +16,113 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+type ReorderOptions = {
+  sortType: SortType;
+  isReversed: boolean;
+};
+
+export function getReorderedGoods(
+  goods: string[],
+  { sortType, isReversed }: ReorderOptions,
+) {
+  const visibleGoods = [...goods];
+
+  visibleGoods.sort((good1, good2) => {
+    switch (sortType) {
+      case SortType.LENGTH:
+        return good1.length - good2.length;
+      case SortType.ALPHABET:
+        return good1.localeCompare(good2);
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [isReversed, setIsReversed] = useState(false);
+  const [sortType, setSortType] = useState(SortType.NONE);
+  const resetButton = sortType !== SortType.NONE || isReversed;
+  const recordedGoods = getReorderedGoods(goodsFromServer, {
+    sortType,
+    isReversed,
+  });
+
+  const handleSort = (sortingType: SortType) => {
+    setSortType(sortingType);
+  };
+
+  const setReverse = () => {
+    setIsReversed(!isReversed);
+  };
+
+  const setReset = () => {
+    setSortType(SortType.NONE);
+    setIsReversed(false);
+  };
```

Comment:
Usually `set` prefix is generally used in state setter. When you define your own function to handle something in response to an event I would propose to use something like:
`handleReverseButton`, `handleReverse`, `handleResetButton`, `handleReset`
It allows to quickly distinguish between function setting the state from `useState` hook and additional custom functions.